File: bucketsofcandy/sol/sol.py
from math import ceil

# numberOfBuckets > numberOfCandies
numberOfBuckets, numberOfCandies = map(int, input().split())
candyArray = [int(input()) for _ in range(numberOfCandies)]

# Binary search to awnser
lowestPossible = 1
highestPossible = max(candyArray)

while lowestPossible != highestPossible:

	# Middle is max amount per bucket for this iteration
	middle = (lowestPossible + highestPossible) // 2

	totalNumberOfBucketsTaken = 0

	for candyAmount in candyArray:
		totalNumberOfBucketsTaken += ceil(candyAmount / middle)

	if totalNumberOfBucketsTaken > numberOfBuckets:
		lowestPossible = middle + 1
	else:
		highestPossible = middle

print(lowestPossible)

File: bucketsofcandy/sol/sol.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
	int numberOfBuckets, numberOfCandies;
	cin >> numberOfBuckets >> numberOfCandies;

	int candyArray[numberOfCandies];
	for (int i = 0; i < numberOfCandies; i++) {
		cin >> candyArray[i];
	}

	int lowestPossible = 1;

	int highestPossible = 0;
	for (int x : candyArray) {
		highestPossible = (x > highestPossible) ? x : highestPossible;
	}

	// Binary search to awnser

	while (lowestPossible != highestPossible) {
		// Middle is max amount per bucket for this iteration
		int middle = (lowestPossible + highestPossible) / 2;

		int totalNumberOfBucketsTaken = 0;

		for (int candyAmt : candyArray) {
			totalNumberOfBucketsTaken += ceil(candyAmt * 1.0 / middle);
		}

		if (totalNumberOfBucketsTaken > numberOfBuckets) {
			lowestPossible = middle + 1;
		} else {
			highestPossible = middle;
		}	
	}

	cout << lowestPossible << endl;

	return 0;
}